openpnp/openpnp
# Feature/blinds feeder named group

# Description
Named group option for BlindsFeeder which acts in addition to fiducial location to determine connected feeders.

# Justification
If multiple tiles of Blindfeeders are used in the same physical location for different jobs/builds the feeders configuration will be confused.
This benefits everyone using BlindsFeeders a significant amount.  Default behavior is by location.

# Instructions for Use

- In the BlindsFeeder tape settings there is a new editable combo box. Either choose an existing group name or enter a new one to override location based grouping.
- Leave the group name as "Default" to get the normal location based behavior.
- Entering an existing group name for a feeder with default fiducial locations or a feeder with no other connected feeders will join it to that group and give it the correct fiducials.
- Each group name can only be used for a single, co-located feeder group. Entering an existing group name for a group of more than one feeders will be rejected.
- A future suggestion for arranging multiple feeder groups on a common tile is TileName:FeederGroupName.  Please do not use the colon delimiter.
- Setting fiducial1 to the same location as an existing group will not automatically join it with the group.  This protects the name group from having uncontrolled feeders join it from co-located feeder groups.
- Renaming a feeder in a group will rename all of the connected feeders.  It will not separate that feeder from the others.
- Some keywords are treated as default.  These include Location, default, none and "".  Entering any of the location keywords results in reset back to "Default"
- A single feeder can leave a group and be assigned back to "Default".  Any feeders connected to it will be disconnected.

# Implementation Details
1. How did you test the change?
Created BlindsFeederTest which performs a series of  BlindsFeeder behavior tests.  The expected fiducials, group name, feeder number and connected feeders are checked for all feeders at every step. 

Basic behavior test including:
  1. Create feeders, set fiducials and check
  2. Connect second feeders using fiducial1
  2. Create third feeder at another location and check that it does not connect to the previous group
  3. Test setting feeder pocket locations operates as expected
  4. Test moving feeder1 to check that conencted feeder2 moves with it.

and a set of group naming tests

1. Create a feeder, set fiducials and change its group name
2. Check that a second feeder set with the same fiducial1 does not join until its group name is set correctly.
3. Check that a feeder at null location can join a group and have its fiducials set correctly
4. Check that a feeder at defined location but no other connected feeders can join a group and be moved to the correct fiducials
5. Check that a group of more than one feeder can't be assigned another existing group name, either at the same or a different location.

Additional tests:
- Feeder group names are persistent through machine.xml
- Feeder group name dropdown contains all of the used feeder group names. Removing all isntances of a group name results in it being excluded from this list when the feeder configuration view is refreshed.


2. Did you follow the [coding style](https://github.com/openpnp/openpnp/wiki/Developers-Guide#coding-style)?  Yes?
3. No changes to `org.openpnp.spi` or `org.openpnp.model` packages 
4. `mvn test` run and passed
